File: WordBooks/app/src/main/java/com/callor/word/persistance/WordDatabase.java
package com.callor.word.persistance;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import com.callor.word.model.WordVO;

import java.util.concurrent.Executor;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/*
이 클래스에서 abstract 키워드의 용도
RoomDatabase 를 상속받았을때 반드시 재 정의해야할 method 가 있는데
그 설정을 무시하고, 필요한 method 만 재 정의 하겠다 라는 선언
 */
@Database(entities = {WordVO.class},
        version = 1,
        exportSchema = false)
public abstract class WordDatabase extends RoomDatabase {

    /*
    이 method 에서 abstract 키워드의 용도
    만약 WordDatabase 클래스를 상속받는 곳에서는 이 method 는
    반드시 재 정의하라
     */
    public abstract WordDao wordDao();

    /*
    프로그램이 작동되는 과정에서 사용된 변수를 저장하는 메모리는
    운영체제, JVM 등에서 최적화를 수행하며 직접 관리를 한다
    volatile 로 지정된 메모리는 최적화에서 제외하고
    성능에 우선을 두는 메모리 관리를 수행하라 라는 의미

    java 에서 volatile 선언
    JVM 메모리가 아닌 안드로이드 OS 가 관리하는 곳에 변수를 저장하라라
    */
    private static volatile WordDatabase DB_CONN;

    // 이 변수 선언문 이후의 method 는
    // thread 환경에서 background 로 실행하라
    public static final ExecutorService databaseExcutor
            = Executors.newFixedThreadPool(3);

    /*
    SingleTone 객체 선언
    프로젝트가 실행되는 과정에서 한번만 객체를 만들고
    그 객체를 재활용 하는 방식
     */
    public static WordDatabase getInstance(final Context context) {
        if(DB_CONN == null) {
            synchronized (WordDatabase.class) {
                if(DB_CONN == null) {
                    DB_CONN = Room.databaseBuilder(
                            context.getApplicationContext(),
                            WordDatabase.class,"wordDB"
                    ).build();
                }
            }
        }
        return DB_CONN;
    }

}

File: WordBook/app/src/main/java/com/callor/word/persistance/WordDatabase.kt
package com.callor.word.persistance

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.sqlite.db.SupportSQLiteDatabase
import com.callor.word.model.Word
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

// 추상클래스
// 인터페이스는 implement 를 하면 인터페이스에 정의된
// 모든 함수(method) 를 반드시 재 정의해야 한다
// 인터페이스를 상속받은 클래스에 abstract 키워드를 부착하면
// 필요한 method 만 재 정의하여 사용할 수 있다
// 일부 함수는 parents 에서 정의하고
// 상속받은 곳에서 그대로 사용하거나, 또는 재정의하여 사용할 수 있다
@Database(entities = arrayOf(Word::class),
    version = 1, exportSchema = false)
abstract class WordDatabase : RoomDatabase() {

    // 추상 메서드
    // 내용 코드가 없는 함수
    // 모양만 정의된 함수
    abstract fun wordDao() : WordDao

    // Java 에서 static 클래스에 해당하는 키워드
    companion object {

        // DB 연결을 위한 커넥션 생성하기 시작
        private var DB_CONN : WordDatabase? = null
        fun getDataBase (
            context : Context,
            scope : CoroutineScope
        ) : WordDatabase {
            return  DB_CONN ?: synchronized(this) {
                // Room.dataBaseBuilder 를 사용하여
                // DBMS 와 연결하는 커넥션 만들기

                val conn = Room.databaseBuilder(
                    context.applicationContext,
                    WordDatabase::class.java,
                    // DBMS 와 연결되는 커넥션의 별명을
                    // WordDB 로 한다
                    "WordDB"
                )
                    .fallbackToDestructiveMigration()
                    .build()

                // 생성한 커넥션을 DB_CONN 에 저장하고
                DB_CONN = conn
                // 밖으로 return 하기
                conn
            }
        } // end getDataBase

    }

}